syncstorage/storage/sql/__init__.py

```diff
@@ -792,7 +798,9 @@ def _purge_expired_batch_items(self, grace_period=0, max_per_loop=1000):
             })
             num_purged += res["num_purged"]
             is_incomplete = is_incomplete or not res["is_complete"]
-        self._maybe_optimize_table("OPTIMIZE_BUI_TABLE", {"bui": table})
+            self._maybe_optimize_table_after_purge("OPTIMIZE_BUI_TABLE", {
+                "bui": table
+            })
```

Comment:
Also there was a bug with the indentation here; previously it would only run after the entire loop had completed, optimizing only the last table in the list instead of optimizing each table in turn :-(